pkg/build/controller/buildpod/factory.go

```diff
@@ -0,0 +1,213 @@
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+)
+
+const (
+	// We must avoid processing build pods until the build and pod stores have synced.
+	// If they haven't synced, to avoid a hot loop, we'll wait this long between checks.
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
+)
+
+// NewBuildPodController creates a new BuildPodController.
+func NewBuildPodController(buildInformer, podInformer cache.SharedIndexInformer, kc kclientset.Interface, oc osclient.Interface) *BuildPodController {
+	buildUpdater := buildclient.NewOSClientBuildClient(oc)
+	buildLister := buildclient.NewOSClientBuildClient(oc)
+	c := &BuildPodController{
+		buildUpdater: buildUpdater,
+		secretClient: kc.Core(),
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.runPolicies = policy.GetAllRunPolicies(buildLister, buildUpdater)
+
+	c.buildStore.Indexer = buildInformer.GetIndexer()
+	c.podStore.Indexer = podInformer.GetIndexer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.updatePod,
+		DeleteFunc: c.deletePod,
+	})
+
+	c.buildStoreSynced = buildInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *BuildPodController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the build store to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down build pod controller")
+	c.queue.ShutDown()
+}
+
+func (bc *BuildPodController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !bc.buildStoreSynced() || !bc.podStoreSynced() {
+		glog.V(4).Infof("Waiting for the build and pod caches to sync before starting the build pod controller worker")
+		select {
+		case <-time.After(storeSyncedPollPeriod):
+		case <-stopCh:
+			return
+		}
+	}
+	close(ready)
+}
+
+func (bc *BuildPodController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildPodController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	pod, err := bc.getPodByKey(key.(string))
+	if err != nil {
+		glog.Error(err.Error())
+	}
+
+	if pod == nil {
+		return false
+	}
+
+	err = bc.HandlePod(pod)
+	bc.handleErr(err, key, pod)
+
+	return false
+}
+
+func (bc *BuildPodController) getPodByKey(key string) (*kapi.Pod, error) {
+	obj, exists, err := bc.podStore.Indexer.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve pod %q from store: %v", key, err)
+		bc.queue.Add(key)
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("Pod %q has been deleted", key)
```

Comment:
yeah ok i was thinking of a different code flow where we get a build event, lookup the buildpod annotation value, and then try to retrieve the build pod by that key.  In *that* case, it's a problem to query a cache for the build pod because it could result in a cache miss and a false result that the pod doesn't exist when it does.

(and the inverse case, where we get a pod event, look up the build associated w/ the pod and again, could have a cache miss if the cache is stale)